### PR TITLE
add solution to task 3 GET/api/categoreis with TDD and error handling

### PR DESCRIPTION
**Pull request by KonaroKam**
Adds code to be able to make GET request to path api/categories and error handling in case of bad paths. TDD Included.

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
+const express = require("express");
+const app = express();
+
+const { getCategories } = require("./controllers/categories_controllers");
+
+app.use(express.json());
+
+app.get("/api/categories", getCategories);
+
+app.all("*", (req, res, next) => {
```
Comment (elijaharley) [inline app.js]: This should be '/api/*'

File: __tests__/app.test.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  test("Get request to /api/categories responds with array of objects", () => {
+    return request(app)
+      .get("/api/categories")
+      .expect(200)
+      .then(({ body }) => {
+        expect(body).toBeInstanceOf(Array);
+        expect(body.length).toBe(4);
+        body.forEach((obj) => {
+          expect(obj).toEqual(
+            expect.objectContaining({
```
Comment (elijaharley) [inline __tests__/app.test.js:37]: Nice use of objectContaining here, and remembering to check length.

File: __tests__/app.test.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+const seed = require("./../db/seeds/seed");
+const db = require("./../db/connection");
+const testData = require("./../db/data/test-data");
+const app = require("../app");
+
+beforeEach(() => seed(testData));
+afterAll(() => db.end());
+
+describe("Error handling Bad Paths", () => {
```
Comment (elijaharley) [inline __tests__/app.test.js:14]: You will only need one 404 test here - we want our code to be as DRY as possible.
make sure the description is referring to the endpoint eg 'Handling Bad Paths on /api route'

File: __tests__/app.test.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
+const { afterAll, beforeEach } = require("@jest/globals");
```
Comment (elijaharley) [inline __tests__/app.test.js:1]: These will have auto imported but don't need to be explicitly required.

File: app.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
+const express = require("express");
+const app = express();
+
+const { getCategories } = require("./controllers/categories_controllers");
+
+app.use(express.json());
```
Comment (elijaharley) [inline app.js:7]: Make sure youonly add code that you need - currently we are not using this - but we will later.

File: app.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+app.all("*", (req, res, next) => {
+  res
+    .status(404)
+    .send({
+      msg: "Resource cannot be found. Check path you are trying to access before trying again.",
+    });
+});
+
+//JS ERROR HANDLING
```
Comment (elijaharley) [inline app.js:19]: Good that you're considering adding these error handlers but be careful about adding specifics such as 22P02 if you don't need them - we may get different error codes from PSQL for different projects.

File: controllers/categories_controllers.js [hunk 7]
```diff
@@ -0,0 +1,12 @@
+const { fetchCategories } = require("../models/categories_models");
+
+exports.getCategories = (req, res, next) => {
+  //stuff
```
Comment (elijaharley) [inline controllers/categories_controllers.js]: Make sure you remove extra comments and console.logs

File: controllers/categories_controllers.js [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+  fetchCategories().then((categories) => {
+        console.log(categories)
+      res.status(200).send(categories);
+    })
+    .catch((err) => {
+      next(err);
+    });
```
Comment (elijaharley) [inline controllers/categories_controllers.js:9]: Nice that you've remembered your catch block for error handling

File: models/categories_models.js [hunk 9]
```diff
@@ -0,0 +1,10 @@
+const db = require("../db/connection");
+const format = require("pg-format");
+
+exports.fetchCategories = () => {
+  //models stuff
+  return db.query(`SELECT * FROM categories`).then(({ rows }) => {
+    console.log(rows ,"<<<<rows")
+    return rows;
+  });
+};
```
Comment (elijaharley) [inline models/categories_models.js:8]: Please see previous comments about unnecessary imports/comments/console.logs

**Review by elijaharley — APPROVED** [review]
Nice job on this Qamar - I've left you some comments about things to consider. Be careful about importing dependencies that you're not using yet, and remember to only add code for error handlers etc that you currently need. Make sure you remove comments and console.logs. Otherwise all working well.

File: __tests__/app.test.js (hunk 2 above)
Comment (KonaroKam) [reply __tests__/app.test.js:37]: Thank you!

**Pull request merged**